### PR TITLE
Remove autoloads of Multipart::UploadedFile and Multipart::Generator

### PR DESCRIPTION
**Pull request by jeremyevans**
With this and the webrick handler autoload removal, rack is autoload
free if you do not require rack itself, and only require the parts
of rack you are using.

Take care to avoid the circular requires when switching from
autoload to require_relative.

### CODE REVIEW

**Comment by ioquatix** [comment]
I am fine with this, but I'd prefer moving the contents of `rack/multipart.rb` into `rack/multipart/whatever.rb` (not literally) and then using require_relative without `defined?` - does it make sense?

**Comment by ioquatix** [comment]
Maybe just moving the constants into `rack/multipart/parser.rb` might be an okay name?

**Comment by jeremyevans** [comment]
> Maybe just moving the constants into `rack/multipart/parser.rb` might be an okay name?

I'll check the constants.  If all are used by only Parser, that makes the most sense to me.

File: lib/rack/multipart.rb [hunk 1]
```diff
@@ -60,9 +32,10 @@ def build_multipart(params, first = true)
...
       end
     end
-
   end
 end
 
 require_relative 'request' unless defined?(Rack::Request)
 require_relative 'multipart/parser'
+require_relative 'multipart/uploaded_file' unless defined?(Rack::Multipart::UploadedFile)
+require_relative 'multipart/generator' unless defined?(Rack::Multipart::Generator)
```
Comment (ioquatix) [inline lib/rack/multipart.rb:41]: Are we able to move these to the top of the file and make them unconditional?

File: lib/rack/multipart/generator.rb [hunk 2]
```diff
@@ -1,6 +1,6 @@
 # frozen_string_literal: true
 
-require_relative '../multipart'
+require_relative '../multipart' unless defined?(Rack::Multipart)
```
Comment (ioquatix) [inline lib/rack/multipart/generator.rb:3]: Can this now become `require_relative 'parser'`?

File: lib/rack/multipart/uploaded_file.rb [hunk 3]
```diff
@@ -2,7 +2,7 @@
 
 require 'tempfile'
 require 'fileutils'
-require_relative '../multipart'
+require_relative '../multipart' unless defined?(Rack::Multipart)
```
Comment (ioquatix) [inline lib/rack/multipart/uploaded_file.rb:5]: Can this now become `require_relative 'parser'`?

**Comment by jeremyevans** [comment]
@ioquatix I think all of your suggestions can cause circular require issues.   The issue is these four files can be required in any order: `rack/multipart`, `rack/multipart/parser`, `rack/multipart/generator`, `rack/multipart/uploaded_file`.  I found that even the previous implementation (before this PR, and before the most recent commit in this PR) could result in circular require warnings, so I pushed another commit to fix that.

I'm not against moving things around when it makes sense to do so, but I don't think moving non-parser parts into `lib/rack/multipart/parser.rb` makes sense.

If you want, open an alternative PR for this, but make sure the following commands do not emit warnings:

```
ruby -w -I lib -r rack/multipart-e ''
ruby -w -I lib -r rack/multipart/parser -e ''
ruby -w -I lib -r rack/multipart/generator -e ''
ruby -w -I lib -r rack/multipart/uploaded_file -e ''
```

**Review by ioquatix — APPROVED** [review]

**Comment by ioquatix** [comment]
I'm happy to merge this work and I'll try to make a follow up PR.

**Comment by ioquatix** [comment]
It looks like our cached gems might be incorrect.

**Comment by jeremyevans** [comment]
> It looks like our cached gems might be incorrect.

I think #1871 may fix that. It hasn't been merged yet.

**Comment by ioquatix** [comment]
I've just merged the CI `bundler-cache` option on main.

**Pull request merged**